cirq-core/cirq/ops/boolean_hamiltonian.py

```diff
@@ -114,6 +121,92 @@ def _decompose_(self):
             hamiltonian_polynomial_list, self._qubit_map, self._theta
         )
 
+    def _has_unitary_(self):
+        return True
+
+    @property
+    def gate(self) -> 'cirq.Gate':
+        return BooleanHamiltonianGate(
+            tuple(self._qubit_map.keys()), self._boolean_strs, self._theta
+        )
+
+
+@value.value_equality
+class BooleanHamiltonianGate(raw_types.Gate):
+    """A gate that represents a Hamiltonian from a set of Boolean functions."""
+
+    def __init__(
+        self,
+        parameter_names: Sequence[str],
+        boolean_strs: Sequence[str],
+        theta: float,
+    ):
+        """Builds a BooleanHamiltonianGate.
+
+        For each element of a sequence of Boolean expressions, the code first transforms it into a
+        polynomial of Pauli Zs that represent that particular expression. Then, we sum all the
+        polynomials, thus making a function that goes from a series to Boolean inputs to an integer
+        that is the number of Boolean expressions that are true.
+
+        For example, if we were using this gate for the unweighted max-cut problem that is typically
+        used to demonstrate the QAOA algorithm, there would be one Boolean expression per edge. Each
+        Boolean expression would be true iff the vertices on that are in different cuts (i.e. it's)
+        an XOR.
+
+        Then, we compute exp(-j * theta * polynomial), which is unitary because the polynomial is
+        Hermitian.
+
+        Args:
+            parameter_names: The names of the inputs to the expressions.
+            boolean_strs: The list of Sympy-parsable Boolean expressions.
+            theta: The evolution time (angle) for the Hamiltonian
+        """
+        self._parameter_names: Sequence[str] = parameter_names
+        self._boolean_strs: Sequence[str] = boolean_strs
+        self._theta: float = theta
+
+    def _qid_shape_(self):
+        return (2,) * len(self._parameter_names)
+
+    def on(self, *qubits) -> 'cirq.Operation':
+        return gate_operation.GateOperation(self, qubits)
```

Comment:
This is already implicitly defined on the `cirq.Gate` class, so not needed here.